rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/core/RMCore.java

```diff
@@ -2265,6 +2273,36 @@ public RMState getState() {
         return state;
     }
 
+    public List<String> getToBeRemovedUnavailableNodesUrls() {
+
+        List<String> unavailableNodesUrl = new LinkedList<>();
+
+        unavailableNodesUrl.addAll(this.allNodes.values()
+                                                .stream()
+                                                .filter(this::isNodeUnavailableForTooLong)
+                                                .map(RMNode::getNodeURL)
+                                                .collect(Collectors.toList()));
+
+        unavailableNodesUrl.addAll(this.deployedNodeSources.entrySet()
+                                                           .stream()
+                                                           .map(Entry::getValue)
+                                                           .map(NodeSource::getDeployingAndLostNodes)
+                                                           .flatMap(list -> list.stream()
+                                                                                .filter(this::isNodeUnavailableForTooLong)
+                                                                                .map(RMDeployingNode::getNodeURL))
+                                                           .collect(Collectors.toList()));
+
+        return unavailableNodesUrl;
+    }
+
+    private boolean isNodeUnavailableForTooLong(RMNode node) {
+        int period = PAResourceManagerProperties.RM_UNAVAILABLE_NODES_MAX_PERIOD.getValueAsInt();
+        int periodInMilliseconds = period * 60 * 1000;
+        long now = System.currentTimeMillis();
+        return (node.getState().equals(NodeState.DOWN) || node.getState().equals(NodeState.LOST)) &&
+               (now - node.getStateChangeTime() > periodInMilliseconds);
```

Comment:
Myabe you can add method to RMNode: `millisSinceStateChange()`.

Reply:
Yes, I gave it a try.